hail/src/main/scala/is/hail/expr/ir/agg/StagedRegionValueAggregator.scala

```diff
@@ -0,0 +1,39 @@
+package is.hail.expr.ir.agg
+
+import is.hail.annotations.{Region, StagedRegionValueBuilder}
+import is.hail.asm4s._
+import is.hail.expr.ir.{EmitMethodBuilder, EmitTriplet}
+import is.hail.expr.types.physical.PType
+
+import scala.language.existentials
+
+case class RVAVariable(triplet: EmitTriplet, t: PType) {
+  def setup: Code[Unit] = triplet.setup
+  def m: Code[Boolean] = triplet.m
+  def v[T]: Code[T] = triplet.value[T]
+}
+
+abstract class StagedRegionValueAggregator {
+  type State <: RVAState
+  def eltArgIdx: Int = 0
+
+  def initOpTypes: Array[PType]
+  def seqOpTypes: Array[PType]
+
+  def resultType: PType
+
+  def createState(mb: EmitMethodBuilder): State
+
+  def initOp(state: State, init: Array[RVAVariable], dummy: Boolean): Code[Unit]
```

Comment:
I was having a hard time with the type signature of this function---wanted to be able to call initOp on a generic RVAState and cast to the correct state for the aggregator, but didn't necessarily want to have to put like 
```
s2 = state.asInstanceOf[State]
```
in every method. So I defined two versions of every method, e.g.
```
def initOp(state: State, init: Array[RVAVariable], dummy: Boolean): Code[Unit]
def initOp(state: RVAState, init: Array[RVAVariable]): Code[Unit] = 
    initOp(state.asInstanceOf[State], init, true)
```
in which the second one is the one that's called externally. The dummy variable was necessary since `State <: RVAState`. Wasn't sure of the right way to do this.